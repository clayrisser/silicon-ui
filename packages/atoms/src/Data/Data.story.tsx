// import React from 'react';
// import { withKnobs, text, select } from '@storybook/addon-knobs';
// import Data from './Data';
// import { StyleSheet, View } from 'react-native';
// import Wrapper from '../../storybook/Wrapper';
// import storiesOf from '../../storybook/storiesOf';
// import themes from '../themes';
// import withThemesProvider from '../../storybook/withThemesProvider';

// storiesOf('Data', module)
//   .addDecorator(withKnobs)
//   .addDecorator(withThemesProvider(themes))
//   .addParameters({
//     // docs: { page: docs },
//     // jest: ['Data.spec.tsx']
//   })
//   .add('with knobs', () => (
//     <Wrapper>
//       <Data
//         backgroundColor={text('backgroundColor', 'secondary')}
//         // borderRadius={number('borderRadius', 0)}
//         // height={200}
//         // width={200}
//         autoContrast={select<'A' | 'AA' | 'AAA'>(
//           'autoContrast',
//           {
//             false: '' as 'A',
//             A: 'A',
//             AA: 'AA',
//             AAA: 'AAA'
//           },
//           'A'
//         )}
//       >
//         data
//       </Data>
//     </Wrapper>
//   ));

// const styles = StyleSheet.create({
//   container: { flex: 1, padding: 16, paddingTop: 30, backgroundColor: '#fff' },
//   head: { height: 40, backgroundColor: '#fff' },
//   text: { margin: 6, backgroundColor: '#fff' }
// });
