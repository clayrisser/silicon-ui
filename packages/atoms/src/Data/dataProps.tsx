import { DetailedHTMLProps, TdHTMLAttributes } from 'react';
import { Table } from 'react-native-table-component';
import {
  BackgroundProps,
  BorderProps,
  ColorProps,
  LayoutProps,
  PositionProps,
  ShadowProps,
  SpaceProps,
  TypographyProps
} from 'styled-system';
import { Theme } from '../themes';

export type DetailedHTMLTdProps = DetailedHTMLProps<
  TdHTMLAttributes<HTMLTableDataCellElement>,
  HTMLTableDataCellElement
>;

export interface StyledTdProps
  extends BackgroundProps,
    BorderProps,
    ColorProps,
    LayoutProps,
    PositionProps,
    ShadowProps,
    SpaceProps,
    TypographyProps {}

export interface NativeTdProps extends Table {}

export interface NativeItemProps extends Omit<Table, 'style'> {}

export interface CustomTdProps {
  autoContrast?: boolean | 'A' | 'AA' | 'AAA';
  theme?: Theme;
  style?: any;
}

export interface TdProps
  extends CustomTdProps,
    NativeTdProps,
    NativeItemProps,
    StyledTdProps {}

export const customTdPropsKeys = new Set(['autoContrast', 'theme', 'style']);

export const nativeTdPropsKeys = new Set<string>(['borderStyle', 'style']);

export const nativeItemPropKeys = new Set([
  'bordered',
  'disabled',
  'error',
  'underline'
]);

export interface SplitProps {
  customTdProps: CustomTdProps;
  nativeTdProps: NativeTdProps;
  nativeItemProps: NativeItemProps;
  styledTdProps: StyledTdProps;
}

export function splitProps(props: TdProps): SplitProps {
  const styledTdProps: { [key: string]: any } = {};
  const customTdProps: { [key: string]: any } = {};
  const nativeTdProps: { [key: string]: any } = {};
  const nativeItemProps: { [key: string]: any } = {};
  Object.entries({ ...props }).forEach(([key, prop]: [string, any]) => {
    if (customTdPropsKeys.has(key)) {
      customTdProps[key] = prop;
    } else if (nativeTdPropsKeys.has(key)) {
      nativeTdProps[key] = prop;
    } else if (nativeItemPropKeys.has(key)) {
      nativeItemProps[key] = prop;
    } else {
      styledTdProps[key] = prop;
    }
  });
  return {
    customTdProps,
    nativeTdProps,
    nativeItemProps,
    styledTdProps
  };
}

export const antiForwardTdPropsKeys = new Set<string>([]);
