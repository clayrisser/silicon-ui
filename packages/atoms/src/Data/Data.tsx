import React, { FC } from 'react';
import styled, { StyledComponent } from '@emotion/styled';
import {
  background,
  border,
  color,
  compose,
  layout,
  position,
  shadow,
  space,
  typography
} from 'styled-system';
import useColor from '../hooks/useColor';
import { TdProps, DetailedHTMLTdProps, splitProps } from './dataProps';

const HTMLTableData: StyledComponent<
  DetailedHTMLTdProps,
  TdProps,
  object
> = styled.td(
  compose(
    background,
    border,
    color,
    layout,
    position,
    shadow,
    space,
    typography
  )
);

const Td: FC<TdProps> = (props: TdProps) => {
  const color = useColor(props);
  const { customTdProps, styledTdProps, nativeItemProps } = splitProps({
    ...props,
    color
  });
  return (
    <HTMLTableData
      {...styledTdProps}
      {...nativeItemProps}
      {...(customTdProps as any)}
    />
  );
};

Td.defaultProps = {
  backgroundColor: 'transparent',
  autoContrast: false,
  fontSize: 2,
  fontWeight: 'body',
  lineHeight: 'body'
};

export default Td;
