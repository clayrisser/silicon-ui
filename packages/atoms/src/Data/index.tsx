import Data from './Data';
import { TdProps } from './dataProps';

export { TdProps };
export default Data;
