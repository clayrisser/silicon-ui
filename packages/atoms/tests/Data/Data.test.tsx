import React from 'react';
import renderer from 'react-test-renderer';
// import { render, fireEvent } from '@testing-library/react';
import Data from '../../src/Data';

describe('<Data />', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Data />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
