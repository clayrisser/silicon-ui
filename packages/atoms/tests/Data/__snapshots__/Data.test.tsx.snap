// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Data /> renders correctly 1`] = `
<td
  className="css-15n1v91"
  fontSize={2}
  fontWeight="body"
/>
`;
